src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,167 @@
+            if (i == index) {
+                value = array[i];
+                selectedIndex = i;
+            }
+        }
+        if (selectedIndex == -1) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds for length " + listLength);
+        }
+        return value;
     }
 
     @Override
     public void set(T value, int index) {
-
+        int updatedIndex = -1;
+        for (int i = 0; i < listLength; i++) {
+            if (i == index) {
+                array[i] = value;
+                updatedIndex = i;
+            }
+        }
+        if (updatedIndex == -1) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds for length " + listLength);
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        int removedIndex = -1;
+        T value = null;
+        for (int i = 0; i < listLength; i++) {
+            if (i == index) {
+                value = array[i];
+                array[i] = null;
+                removedIndex = i;
+            }
+        }
+        if (removedIndex == -1) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds for length " + listLength);
+        }
+        for (int i = removedIndex + 1; i < listLength; i++) {
+            array[i - 1] = array[i];
+        }
+        array[listLength - 1] = null;
+        listLength--;
+        return value;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int removedIndex = -1;
+        T value = null;
+        for (int i = 0; i < listLength; i++) {
+            if (array[i] != null && array[i].equals(element)
+                    || (array[i] == element)) {
+                value = array[i];
+                array[i] = null;
+                removedIndex = i;
+            }
+        }
+        if (removedIndex == -1) {
+            throw new NoSuchElementException();
+        }
+        for (int i = removedIndex + 1; i < listLength; i++) {
+            array[i - 1] = array[i];
+        }
+        array[listLength - 1] = null;
+        listLength--;
+        return value;
     }
 
     @Override
     public int size() {
-        return 0;
+        return listLength;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return listLength == 0;
+    }
+
+    private T[] grow(T[] originalArray) {
+        int newSize = (int) Math.ceil(originalArray.length * 1.5);
+        T[] newArray = (T[]) new Object[newSize];
+        System.arraycopy(originalArray, 0, newArray, 0, originalArray.length);
+        return newArray;
+    }
+
+    private int indexOf(Object o) {
+        for (int i = 0; i < array.length; i++) {
+            if (array[i] == o || (array[i] != null && array[i].equals(o))) {
+                return i;
+            }
+        }
+        return -1;
```

Comment:
The 'indexOf' method can be misleading because it returns the index of the first occurrence of the specified element or -1 if the element is not present. However, the method name does not clearly indicate this behavior. Consider renaming it to more accurately reflect its purpose, such as 'findFirstIndexOf'.